zkevm-circuits/src/evm_circuit/execution/create.rs

```diff
@@ -0,0 +1,915 @@
+        let salt = if is_create2 {
+            block.get_rws(step, 7).stack_value()
+        } else {
+            U256::zero()
+        };
+
+        let rw_offset = if is_create2 { 8 } else { 7 };
+        let (caller_balance, _) = block.get_rws(step, rw_offset + 1).account_value_pair();
+        let (caller_nonce, _) = block.get_rws(step, rw_offset + 2).account_value_pair();
+        let is_precheck_ok =
+            if call.depth < 1025 && caller_balance >= value && caller_nonce.as_u64() < u64::MAX {
+                1
+            } else {
+                0
+            };
+
+        self.caller_balance
+            .assign(region, offset, Some(caller_balance.to_le_bytes()))?;
+
+        let (callee_prev_code_hash, was_warm) = if is_precheck_ok == 1 {
+            let (_, was_warm) = block
+                .get_rws(step, rw_offset + 4)
+                .tx_access_list_value_pair();
+            let (callee_prev_code_hash, _) =
+                block.get_rws(step, rw_offset + 5).account_value_pair();
+            (callee_prev_code_hash, was_warm)
+        } else {
+            (U256::from(0), false)
+        };
+
+        // 3 RWs while is_precheck_ok is true
+        // account_write(caller), tx_access_list_write(callee) and  account_read(callee)
+        let [callee_rw_counter_end_of_reversion, callee_is_persistent] = [
+            rw_offset + 11 - (1 - is_precheck_ok) * 3,
+            rw_offset + 12 - (1 - is_precheck_ok) * 3,
+        ]
+        .map(|i| block.get_rws(step, i).call_context_value());
+
+        // retrieve code_hash for creating address
+        let is_address_collision = !callee_prev_code_hash.is_zero();
+        let code_hash_previous_rlc = if is_address_collision {
+            region.code_hash(callee_prev_code_hash)
+        } else {
+            Value::known(F::ZERO)
+        };
+        self.prev_code_hash
+            .assign(region, offset, code_hash_previous_rlc)?;
+        self.not_address_collision.assign(
+            region,
+            offset,
+            F::from((is_address_collision).into()),
+        )?;
+
+        // gas cost of memory expansion
+        let init_code_address =
+            self.init_code
+                .assign(region, offset, init_code_start, init_code_length)?;
+
+        let (_, memory_expansion_gas_cost) = self.memory_expansion.assign(
+            region,
+            offset,
+            step.memory_word_size(),
+            [init_code_address],
+        )?;
+        let (init_code_word_size, _) = self.init_code_word_size.assign(
+            region,
+            offset,
+            (31u64 + init_code_length.as_u64()).into(),
+        )?;
+        let gas_left = step.gas_left.0
+            - GasCost::CREATE.as_u64()
+            - memory_expansion_gas_cost
+            - if is_create2 {
+                u64::try_from(init_code_word_size).unwrap() * GasCost::COPY_SHA3.as_u64()
+            } else {
+                0
+            };
```

Comment:
For shanghai geth traces, it may be fixed to:
```
- u64::try_from(init_code_word_size).unwrap() * if is_create2 {
    INIT_CODE_WORD_GAS + GasCost::COPY_SHA3.as_u64()
} else {
    INIT_CODE_WORD_GAS
};
```